modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/H2IndexCostedBase.java

```diff
@@ -771,18 +790,28 @@ private long rowCost(
                         rowCount = getColumnSize(colStats, rowCount, equalNull);
 
                         if (colStats != null && equalNull == Boolean.TRUE) {
-                            rowsCost = Math.min(5 + Math.max(rowsCost * colStats.nulls() / 100, 1), rowsCost -
+                            rowsCost = Math.min(5 + Math.max(Math.round(rowsCost * nulls(colStats)), 1), rowsCost -
                                     (i > 0 ? 1 : 0));
                             continue;
                         }
                         if (colStats != null && equalNull == Boolean.FALSE)
-                            rowsCost = rowsCost * (100 - colStats.nulls()) / 100;
-
-                        int cardinality = getColumnCardinality(colStats, column);
+                            rowsCost = Math.max(Math.round(rowsCost * (1 - nulls(colStats))), 1);
 
-                        totalCardinality = 100 - ((100 - totalCardinality) * (100 - cardinality) / 100);
-
-                        long distinctRows = Math.round((double) rowCount * totalCardinality / 100);
+                        long distinctRows;
+                        if (colStats == null) {
+                            double cardinality = (double)column.getSelectivity() / 100;
+                            totalCardinality = 1 - (1 - totalCardinality) * (1 - cardinality);
+                            distinctRows = Math.round((double) rowCount * totalCardinality);
+                        }
+                        else {
+                            double cardinality;
+                            if (colStats.total() - colStats.nulls() == 0)
```

Comment:
may be better to extract variable for notNulls as colStats.total() - colStats.nulls()